### PR TITLE
WIP/RFC: Converted WeakKeyDict to hash & compare with object-id

### PR DESCRIPTION
**Pull request by mauro3**
This implements Stefan's suggestion to use object-id hash in the WeakKeyDict:  https://github.com/JuliaLang/julia/issues/24941#issuecomment-404263168, which has currently a "Triage" label.  Maybe this PR can help inform that decision. The changes were pretty minimal and no breakage occurred.

TODO if we go ahead with this:
- [ ] run package evaluator(?)
- [ ] rename `WeakKeyDict` to `WeakKeyIdDict`
- [ ] deprecate
- [ ] NEWS item

I can do this on Friday if triage is positive.

TODO if we don't go ahead with this:
- add a test for key-conversion and the key-dropping (#24941)

x-refs: #3002, #24941, #28182

### CODE REVIEW

File: test/dict.jl [hunk 1]
```diff
@@ -830,6 +830,21 @@ Dict(1 => rand(2,3), 'c' => "asdf") # just make sure this does not trigger a dep
 
     @test_throws ArgumentError WeakKeyDict([1, 2, 3])
 
+    # WeakKeyDict does not convert keys
+    @test_throws ArgumentError WeakKeyDict{Int,Any}(5.0=>1)
+
+    # WeakKeyDict hashes with object-id
+    AA = copy(A)
```
Comment (vtjnash) [inline test/dict.jl]: ``julia
GC.@preserve A AA begin
...
end
```

Comment (mauro3) [reply test/dict.jl]: Thanks, fixed! Quesion: why would they need to be preserved?  The do not go out of scope and thus should not be GC'ed, no?

Comment (StefanKarpinski) [reply test/dict.jl]: `A` is used subsequently so is relatively safe but `AA` isn't so it could be garbage collected.

**Comment by mauro3** [comment]
The `==` is actually needed: https://github.com/JuliaLang/julia/issues/24941#issuecomment-405987699.

**Pull request closed without merging**